### PR TITLE
Parse playlists and identify them by loaded segments.

### PR DESCRIPTION
**Pull request by i-zolotarenko**

### CODE REVIEW

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
+import {
+  Parser,
+  type PlaylistManifest,
+  type MasterManifest,
+  type Segment as ParserSegment,
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:5]: No need for type specification here

**Review by mrlika — COMMENTED** [review]
> This branch has conflicts that must be resolved

Please resolve conflicts

**Review by mrlika — COMMENTED** [review]
`p2p-media-loader-demo` has compile errors

File: packages/p2p-media-loader-hlsjs/src/services/fragment-loader.ts [hunk 2]
```diff
@@ -28,7 +31,22 @@ export class FragmentLoaderBase implements Loader<FragmentLoaderContext> {
...
+      onSuccess: (response, stats, context, networkDetails) => {
+        // console.log("fragment: ", response.url);
+        const playlist =
+          this.segmentManager.videoPlaylists?.getPlaylistBySegmentUrl(
+            response.url
+          ) ??
+          this.segmentManager.audioPlaylists?.getPlaylistBySegmentUrl(
+            response.url
+          );
+        // console.log(playlist);
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/fragment-loader.ts:45]: Don't commit comments

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  getPlaylist(playlistUrl: string): Playlist | null {
+    return this.playlistMap.get(playlistUrl) ?? null;
+  }
+
+  getPlaylistBySegmentUrl(segmentUrl: string): Playlist | null {
+    for (const playlist of this.playlists) {
+      if (playlist.segmentsMap.has(segmentUrl)) return playlist;
+    }
+    return null;
+  }
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:49]: `undefined` is usually better than `null`. For example, the requirement for developers who create TypeScript language: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  type: SegmentType;
+
+  constructor(type: SegmentType, uri: string, playlistUrl: string) {
+    this.type = type;
+    this.uri = uri;
+    this.url = new URL(uri, playlistUrl).toString();
+  }
+}
+
+class TypeGuard {
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:127]: No need in TypeGuard class to use it as  a namespace

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+} from "m3u8-parser";
+
+export class SegmentManager {
+  videoPlaylists: VideoPlaylistsContainer = new VideoPlaylistsContainer();
+  audioPlaylists: AudioPlaylistsContainer = new AudioPlaylistsContainer();
+
+  processPlaylist(content: string, requestUrl: string, responseUrl: string) {
+    const parser = new Parser();
+    parser.push(content);
+    parser.end();
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:15]: Would you happen to know how errors are handled?

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
+import {
+  Parser,
+  type PlaylistManifest,
+  type MasterManifest,
+  type Segment as ParserSegment,
+} from "m3u8-parser";
+
+export class SegmentManager {
+  videoPlaylists: VideoPlaylistsContainer = new VideoPlaylistsContainer();
+  audioPlaylists: AudioPlaylistsContainer = new AudioPlaylistsContainer();
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:10]: Consider refactoring (simplify) to a map that contains both video and audio playlists: `playlists: Map<string, Playlist>` without abstract classes and inheritance

**Review by mrlika — CHANGES_REQUESTED** [review]

**Review by mrlika — CHANGES_REQUESTED** [review]

File: packages/p2p-media-loader-hlsjs/src/services/playlist.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    type: SegmentType;
+    url: string;
+    manifestUrl?: string;
+    mediaSequence: number;
+    index: number;
+  }) {
+    this.type = type;
+    this.index = index;
+    this.url = new URL(url, manifestUrl).toString();
+    this.id = manifestUrl ? `${manifestUrl}-${type}-V${index}` : this.url;
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/playlist.ts:27]: Use the base part of `manifestUrl` and `this.url` without `?`

**Review by mrlika — CHANGES_REQUESTED** [review]

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        const playlist = new Playlist({
+          type: "unknown",
+          url: responseUrl,
+          manifestUrl: this.manifestUrl,
+          mediaSequence,
+          index: -1,
+        });
+        playlist.setSegments(segments);
+        this.playlists.set(responseUrl, playlist);
+      }
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:54]:
```suggestion
      let playlist = this.playlists.get(responseUrl);

      if (!playlist && !this.manifest) {
        playlist = new Playlist({
          type: "unknown",
          url: responseUrl,
          manifestUrl: this.manifestUrl,
          mediaSequence,
          index: -1,
        });
        this.playlists.set(responseUrl, playlist);
        }
      }
      
      if (playlist) playlist.setSegments(segments);
```

**Review by mrlika — CHANGES_REQUESTED** [review]

File: packages/p2p-media-loader-hlsjs/src/services/playlist.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
+import { Segment as ParserSegment } from "m3u8-parser";
```
[the remark this answers is not in the document]
Comment (i-zolotarenko) [reply packages/p2p-media-loader-hlsjs/src/services/playlist.ts:1]: Actually I have the Segment class in this file so I have to rename this import.

File: packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import { MasterManifest, Parser } from "m3u8-parser";
```
[the remark this answers is not in the document]
Comment (i-zolotarenko) [reply packages/p2p-media-loader-hlsjs/src/services/segment-mananger.ts:1]: Can I investigate it later? I guess its behavior probably depends on certain player.

**Review by mrlika — APPROVED** [review]

File: packages/p2p-media-loader-hlsjs/src/services/playlist.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    this.mediaSequence = mediaSequence;
+  }
+
+  setSegments(segments: ParserSegment[]) {
+    const mapEntries = segments.map<[string, Segment]>((s) => {
+      const segment = new Segment(s.uri, this.url, s.byterange);
+      return [segment.localId, segment];
+    });
+    this.segmentsMap = new Map(mapEntries);
+  }
```
Comment (mrlika) [inline packages/p2p-media-loader-hlsjs/src/services/playlist.ts:41]:
```suggestion
    this.type = type;
    this.index = index;
    this.url = new URL(url, manifestUrl?.response).toString();
    this.id = manifestUrl?.request
      ? `${getUrlWithoutParameters(manifestUrl.request)}-${type}-V${index}`
      : getUrlWithoutParameters(this.url);
    this.mediaSequence = mediaSequence;
  }

  setSegments(segments: ParserSegment[]) {
    const mapEntries = segments.map<[string, Segment]>((s) => {
      const segment = new Segment(s.uri, this.url, s.byterange);
      return [segment.localId, segment];
    });
    this.segmentsMap = new Map(mapEntries);
  }
```

**Review by mrlika — CHANGES_REQUESTED** [review]

**Review by mrlika — APPROVED** [review]

**Pull request merged**